sklearn/linear_model/sag_fast.pyx

```diff
@@ -97,84 +292,131 @@ def sag(SequentialDataset dataset,
+    cdef double* gradient = <double*> gradient_array.data
+
     # the scalar used for multiplying z
     cdef double wscale = 1.0
 
-    # helpers
-    cdef double val, cum_sum, gradient_change
-    cdef int j
-
     # the cumulative sums for each iteration for the sparse implementation
     cumulative_sums[0] = 0.0
 
+    # Loss function to optimize
+    cdef LossFunction loss
+    # Wether the loss function is multinomial
+    cdef bint multinomial = False
+    # Multinomial loss function
+    cdef MultinomialLogLoss multiloss
+
+    if loss_function == "multinomial":
+        multinomial = True
+        multiloss = MultinomialLogLoss()
+    elif loss_function == "log":
+        loss = Log()
+    elif loss_function == "squared":
+        loss = SquaredLoss()
+    else:
+        raise ValueError("Invalid loss parameter: got %s instead of "
+                         "one of ('log', 'squared', 'multinomial')"
+                         % loss_function)
+
     with nogil:
         start_time = time(NULL)
         while True:
+            if infinity:
+                break
             for itr in range(n_samples):
 
                 # extract a random sample
-                current_index = dataset.random(&x_data_ptr,
-                                               &x_ind_ptr,
-                                               &xnnz,
-                                               &y,
-                                               &sample_weight)
+                sample_index = dataset.random(&x_data_ptr, &x_ind_ptr, &xnnz,
+                                              &y, &sample_weight)
 
                 # update the number of samples seen and the seen array
-                if seen[current_index] == 0:
+                if seen[sample_index] == 0:
                     num_seen += 1
-                    seen[current_index] = 1
+                    seen[sample_index] = 1
 
                 # make the weight updates
                 if itr > 0:
                     for j in range(xnnz):
-                        idx = x_ind_ptr[j]
+                        f = x_ind_ptr[j]
+                        idx = f * n_classes - 1
 
                         cum_sum = cumulative_sums[itr - 1]
-                        if feature_hist[idx] != 0:
-                            cum_sum -= cumulative_sums[feature_hist[idx] - 1]
+                        if feature_hist[f] != 0:
+                            cum_sum -= cumulative_sums[feature_hist[f] - 1]
 
-                        weights[idx] -= cum_sum * sum_gradient[idx]
+                        for c in range(n_classes):
+                            idx += 1  # idx = f * n_classes + c
+                            weights[idx] -= cum_sum * sum_gradient[idx]
 
-                        feature_hist[idx] = itr
+                            # check to see that the weight is not inf or NaN
+                            if not skl_isfinite(weights[idx]):
+                                infinity = True
```

Comment:
Would it be better to raise the ValueError here and do away with the infinity variable instead of breaking multiple times and check at the end?

There are only two times where the value of `infinity` changes.